### ISSUE TITLE
MissingMethodException for dynamically loaded Assembly

### ISSUE THREAD

**Issue by matthid** [dotnet/coreclr#7043]
In my situation I'm trying to compile an application at runtime and loading it into a custom `AssemblyContext` as suggested by another issue as workaround: https://github.com/dotnet/coreclr/issues/6411

With this workaround in place I noticed another error. The runtime doesn't seem to load a required `System.IO.Compression.ZipFile` assembly. Instead it fails when using the `ZipArchive` extension methods defined in `ZipFilExtensions`.

I hooked into the `AssemblyResolve` event and the following is what I'm seeing:

```
Starting Target: DotnetCoreCreateZipPackages (==> BootstrapAndBuildDnc)
Trying to resolve: System.IO.Compression, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
Redirect assembly load to known assembly: System.IO.Compression, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 (Some
  "C:\PROJ\FAKE\nuget\dotnetcore\Fake.netcore\current\System.IO.Compression.dll")
Finished Target: DotnetCoreCreateZipPackages
Trying to resolve: System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
Unable to find assembly {FullName =
  "System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089";
 Version = "4.0.0.0";
 Location =
  "C:\PROJ\FAKE\.fake\build.fsx\packages\Microsoft.NETCore.Portable.Compatibility\ref\netstandard1.0\System.Core.dll";}. (Error: System.BadImageFormatException: Could not load file or assembly 'System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089, processorArchitecture=MSIL'. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
File name: 'System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089, processorArchitecture=MSIL' ---> System.BadImageFormatException: Cannot load a reference assembly for execution.
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at Fake.Runtime.CoreCache.realLoadAssembly@205(AssemblyLoadContext loadContext, Boolean printDetails, AssemblyInfo assemInfo)
   at Fake.Runtime.CoreCache.loadAssembly(AssemblyLoadContext loadContext, Boolean printDetails, AssemblyInfo assemInfo)) Could not resolve: System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
Error:
System.MissingMethodException: Method not found: 'System.IO.Compression.ZipArchiveEntry System.IO.Compression.ZipFileExtensions.CreateEntryFromFile(System.IO.Compression.ZipArchive, System.String, System.String)'.
   at Fake.IO.Zip.createZipP[a](String fileName, a comment, Int32 level, IEnumerable`1 items)
   at Fake.IO.Zip.CreateZip[a](String workingDir, String fileName, a comment, Int32 level, Boolean flatten, IEnumerable`1 files)
   at FSI_0002.clo@587-37.Invoke(Unit _arg2)
   at Fake.Core.Targets.runSingleTarget(TargetTemplate`1 target)
```

So it is not even trying to load the required assembly. Interestingly when [I add](https://github.com/matthid/FAKE/commit/80c3d2da3eaacf0c119423b0e46e4922653b6670)

``` fsharp
    let lc = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(typeof<MyClass>.GetTypeInfo().Assembly)
    let n = AssemblyName "System.IO.Compression.ZipFile"
    lc.LoadFromAssemblyName(n) |> ignore
```

just before using the failing method everything works as it should...

You can reproduce this by typing the following into a git bash (windows 10 x64):

``` bash
git clone git@github.com:matthid/FAKE.git
export APPDATA=$HOME LocalAppData=$HOME # Only when on unix
cd FAKE
git checkout 8a59f34
SkipTests=true NO_DOTNETCORE_BOOTSTRAP=true ./build.sh # Builds all the stuff we need, will take a while
rm nuget/dotnetcore/Fake.netcore/fake-dotnetcore*.zip
COREHOST_TRACE=1 ./nuget/dotnetcore/Fake.netcore/current/Fake.netcore.exe -v run build.fsx -t DotnetCoreCreateZipPackages --singletarget # NOTE Works first time
rm nuget/dotnetcore/Fake.netcore/fake-dotnetcore*.zip
COREHOST_TRACE=1 ./nuget/dotnetcore/Fake.netcore/current/Fake.netcore.exe -v run build.fsx -t DotnetCoreCreateZipPackages --singletarget # Now fails every run, I think in the uncached run the required assembly is loaded in another path
```

To add the [workaround](https://github.com/matthid/FAKE/commit/80c3d2da3eaacf0c119423b0e46e4922653b6670)

``` bash
git merge 80c3d2d
rm -rf .fake/build.fsx # just to be save 2nd time
rm -rf ~/.nuget/packages/Fake.* # when running second time, just to be safe
SkipTests=true NO_DOTNETCORE_BOOTSTRAP=true ./build.sh # Builds all the stuff we need, will take a while
rm nuget/dotnetcore/Fake.netcore/fake-dotnetcore*.zip
COREHOST_TRACE=1 ./nuget/dotnetcore/Fake.netcore/current/Fake.netcore.exe -v run build.fsx -t DotnetCoreCreateZipPackages --singletarget # NOTE Works first time
rm nuget/dotnetcore/Fake.netcore/fake-dotnetcore*.zip
COREHOST_TRACE=1 ./nuget/dotnetcore/Fake.netcore/current/Fake.netcore.exe -v run build.fsx -t DotnetCoreCreateZipPackages --singletarget # NOTE Works second time
```

This uses the latest RTM packages of the dotnetcore runtime. I couldn't reproduce this on a smaller sample but I tried to make it as easy as possible to reproduce.

I noticed that AssemblyResolve is not triggered for ZipFile in the first (working) run either. Any help is welcome. Maybe related https://github.com/dotnet/corefx/issues/10787

Please ask if you have further questions about this setup.

TLDR: Runtime is throwing `MissingMethodException` without even trying to load the required assembly. When loading the assembly by hand it works.

**Comment by christianhuening** [comment]
Hi there,

I think I might have the same problem. I am dynamically loading some sort of plugins during the startup of my application. When a plugin is standalone (does not reference any other type from a different assembly), Activator.CreateInstance() just works fine. But when the plugin (A) references types from another assembly B (and so forth), Activator.CreateInstance() fails with MissingMethodException for the ctors of those types etc.

**Comment by gkhanna79** [comment]
@matthid The exception details you pasted above show the following:

`System.BadImageFormatException: Cannot load a reference assembly for execution.`

Looks like you are trying to load a reference assembly as opposed to the implementation assembly. Also, LoadFromAssemblyPath will only trigger Resolving event if the assembly being loaded has dependencies not yet loaded in the load context. 

Are you not able to LoadFromAssemblyPath the actual implementation assembly?

**Comment by gkhanna79** [comment]
> But when the plugin (A) references types from another assembly B (and so forth), Activator.CreateInstance() fails with MissingMethodException for the ctors of those types etc.

@christianhuening When the above happens, is assembly B already part of the app?

**Comment by christianhuening** [comment]
@gkhanna79 Well the assemblies are not a part of the app so far, if you mean at compile time. They are only referenced by the assembly from which I try to instantiate a type.

Essentially at one point I am using a custom AssemblyLoader to search for the types in my DLLs in a certain directory and then I just call Activator.CreateInstance(...) on the types I want/need (i.e. that implement some specific interface). It's a plugin mechanism that I solved with Mono.Addins in .NET Framework / Mono times.

After reflecting over the assemblies and looking for the needed types, they do not show up in DependencyContext.Default.CompileLibraries. I guess this is expected, but I have to admit that I did not really manage to understand the new system already yet. The document is a bit thin on it and I didn't find a writeup so far.

So is there something I need to do to "add" an assembly to the app in order for the Activator to be able to find it?

**Comment by gkhanna79** [comment]
> So is there something I need to do to "add" an assembly to the app in order for the Activator to be able to find it?

For an assembly to be part of the app, it needs to be built as part of the app and end up having its reference in the app's deps.json.

For the scenario you are working with (plugins), the approach should be what you are already using:

1) Create custom ALC and load an assembly A within it.
2) A's load could trigger load of its dependencies that are not part of the app by default. To handle such dependencies, ensure you have wired-up to Resolving event of the ALC instance and then load the dependency in the load context.

**Comment by matthid** [comment]
> @matthid The exception details you pasted above show the following:
>
>System.BadImageFormatException: Cannot load a reference assembly for execution.
>
>Looks like you are trying to load a reference assembly as opposed to the implementation assembly. >Also, LoadFromAssemblyPath will only trigger Resolving event if the assembly being loaded has >dependencies not yet loaded in the load context.
>
>Are you not able to LoadFromAssemblyPath the actual implementation assembly?

I think that error (or rather warning in my case) is unrelated. I could be wrong but the workaround strongly suggests that it is. I think the `System.Core` assembly is required on the error path only (ie when the exception is already thrown).
When I try to fix this warning the runtime crashes, see https://github.com/dotnet/coreclr/issues/7044


Regarding the side discussion: Yes we talk about dynamically loading assemblies at runtime into a custom loader context.

**Comment by christianhuening** [comment]
> 2) A's load could trigger load of its dependencies that are not part of the app by default. To handle such dependencies, ensure you have wired-up to Resolving event of the ALC instance and then load the dependency in the load context.

Ok, so I am required to keep the ALC instance along while trying to instantiate stuff? Currently I am just using the ALC to get to the Type objects and then continue working with only the Type objects.

**Comment by gkhanna79** [comment]
> Ok, so I am required to keep the ALC instance along while trying to instantiate stuff? 

Yes, since attempt to get a Type may require creating the type and thus, loading the assembly in which the type is defined (and any other assemblies defining the types that are dependencies of the type being created).

**Comment by gkhanna79** [comment]
@christianhuening @matthid Can this be closed?

**Comment by christianhuening** [comment]
@gkhanna79 Well my stuff is working now, so this may be closed. However I'd be glad to have better documentation for the reflection stuff once it's considerably stable.

**Comment by matthid** [comment]
@gkhanna79 I'd say the bug is still there. I'm closing this as it is non-trivial to reproduce on the current version (preview4) without converting all projects. `dotnet migrate` won't work as F# is not supported so it is a lot of work....

I will re-open a new issue when I can reproduce again.

**Comment by matthid** [comment]
I upgraded to the latest tooling and runtime and this issue is back again.
Now even the workaround is no longer working :/
https://ci.appveyor.com/project/SteffenForkmann/fake/build/1.0.3038

**Comment by gkhanna79** [comment]
This is strange is there were no loader fixes/changes in the latest build. Can you please try to create a reduced repro with latest (1.1.1) .NET Core and share it out?

**Comment by matthid** [comment]
Ok for anyone stumbling onto this: Sadly I couldn't reproduce this in a standalone repro. But I could verify the behavior against 1.1.1 (I downloaded the runtime and used `dotnet <scd>.dll`).

I think this problem went away after I did the following:
 - read https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md
 - I noticed that I needlessly used the Resolve event instead of just loading the first assembly in the new LoadContext as well
 - I noticed that I still had a "hack" active from the early days which falls back to assemblies from the SCD-package (AppPath). Since there is an automatic fallback when you use LoadContext I removed this part. This was probably from the early "AppDomain" thinking.

As I probably used the API in a wrong way (I still think the error shouldn't have happened) and now it seems to work I will close this finally :)

Thanks @gkhanna79

**Comment by gkhanna79** [comment]
Glad to know you got it resolved :)
